feat(docs): added accessibility guide (#2122)
lucide-icons/lucide: docs/guide/advanced/accessibility.md

element. In most cases they will be part of a badge, button (including icon buttons), navigation
item or other interactive UI element.

::: warning
In case some of your icons stand alone, and they serve a non-decorative function, make sure to provide
the appropriate accessible label for them via the `aria-label` attribute.
:::

![](../../images/a11y/alttext-standalone.svg?raw=true)

In general try to avoid using functional icons with no interactivity, we recommend that:

1) you either add a visible description next to them, or
2) place them in badges, labels or on buttons, and at least add a tooltip to them.

In any such case, it is preferred that the accessible name be provided for these interactive
elements (badges, buttons, nav items etc.) only, _not_ the icons themselves.

### On buttons

Do not provide an accessible label to icons when used on a button, as this label will be read out by
screen readers, leading to nonsensical text.

![](../../images/a11y/alttext-buttons.svg?raw=true)

### On icon buttons

Icon buttons are buttons that do not contain any visible text apart from the icon itself (think of
the close button of a dialog for example).

> ![](../../images/a11y/alttext-iconbuttons.svg?raw=true)

Reviewer: Why?

readers.

### On standalone icons

Icons are usually very unlikely to stand on their own with no semantically meaningful wrapper
element. In most cases they will be part of a badge, button (including icon buttons), navigation
item or other interactive UI element.

::: warning
In case some of your icons stand alone, and they serve a non-decorative function, make sure to
provide the appropriate accessible label for them.
:::

![In short: provide accessible label for semantic icons, but not for decorative icons.](../../images/a11y/alttext-standalone.svg?raw=true)

In general try to avoid using functional icons with no interactivity, we recommend that:

1) you either add a visible description next to them, or
2) place them in badges, labels or on buttons, and at least add a tooltip to them.

In any such case, it is preferred that the accessible name be provided for these interactive
elements (badges, buttons, nav items etc.) only, _not_ the icons themselves.

### On buttons

Do not provide an accessible label to icons when used on a button, as this label will be read out by
screen readers, leading to nonsensical text.

![](../../images/a11y/alttext-buttons.svg?raw=true)

> 

Author: @ericfennis: These could be useful, but they break up the guide quite a bit, I've hidden them within details elements.

> 

Reviewer 2: Nice looks great!

### On icon buttons

Icon buttons are buttons that do not contain any visible text apart from the icon itself (think of
the close button of a dialog for example).

As previously stated, you should provide your accessible label on the icon button itself, not the
contained icon.

![](../../images/a11y/alttext-iconbuttons.svg?raw=true)

## A note on `aria-label`

Although you could provide accessible labels to your elements via the `aria-label` attribute, we
generally recommend against this for and instead suggest that

> you [use your chosen CSS framework's "visually hidden" utility whenever possible](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).

Reviewer 2: Can you explain why we should be "against" this? It would be nice if we explained this more in the text, about why using "visually hidden" is better than using `aria-label`, we also could provide some sources for blog posts or documentation to prove our "recommendation".

Also maybe this reads a bit nicer
``` suggestion
generally recommend visually hiding textual descriptions. For example by [using utility CSS framework's "visually hidden"](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
```

Author: That means something different though. Also, the link itself is the resource that explains why `.visually-hidden` > `[aria-label]`, I've rephrased the link text so that it's clearer.

```suggestion
generally recommend against this and instead suggest that you use your chosen CSS framework's "visually hidden" utility whenever possible. You can [read more about why `aria-label` might not be the best solution](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).
```

Reviewer 2: I'm still not happy with the phraise "we generally recommend against this". That suggest that we are against the use of `aria-label`. And from what I found there nothing wrong with it.
I think you should approach it more positively. "If you are using a CSS framework like Tailwind, we recommend using visually hidden utilities whenever possible"

Reviewer: The options from best to worst:

1. visible label
2. visually hidden label
3. aria-label

Author: > And from what I found there nothing wrong with it.

Well, there kind of is, that's what the article linked is about: https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/#the-aria-label-attribute-is-not-always-translated

But sure, we can rephrase it to be more lenient.

Reviewer 2: They use the word "avoid" instead of "against", which is I think more right fit. Since using `aria-label` is not wrong. I  think we also should link this article to this paragraph, it gives a bit more context to this topic.

Author: No, in certain scenarios it might actually be wrong, since `aria-label` will generally be ignored on non-interactive elements. I don't believe linking to a certain paragraph is necessary, since the entirety of article is about why `aria-label` should be avoided if possible, there isn't a single reason, there are multiple.

Since this thread is on an outdated commit, I've made a suggestion here: https://github.com/lucide-icons/lucide/pull/2122/files#r1636386974

Reviewer 2: Ah ok true, tnx for suggestion think it reads a bit nicer as well.

![](../../images/a11y/alttext-buttons.svg?raw=true)

::: details Code examples

```tsx
// Don't do this
<button>
  <Plus aria-label="Plus icon"/>
  Add document
</button>

// Do this, just leave it
<button>
  <Plus/>
  Add document
</button>
```

:::

### On icon buttons

Icon buttons are buttons that do not contain any visible text apart from the icon itself (think of
the close button of a dialog for example).

As previously stated, you should provide your accessible label on the icon button itself, not the
contained icon.

![](../../images/a11y/alttext-iconbuttons.svg?raw=true)

::: details Code examples

```tsx
// Don't do this
<button class="btn-icon">
  <Home/>
</button>

// Don't do this either
<button class="btn-icon">
  <Home aria-label="Home icon"/>
</button>

// This works but might not be the best solution, see below
<button aria-label="Go to home" class="btn-icon">
  <Home/>
</button>

// Do this instead
<button class="btn-icon">
  <Home/>
  <span class="visually-hidden">Go to home</span>
</button>
```

:::

## A note on `aria-label`

Although you could provide accessible labels to your elements via the `aria-label` attribute, we

> generally recommend against this and instead suggest that you use your chosen CSS framework's "

Author: ```suggestion
generally recommend avoiding this and instead suggest that you use your chosen CSS framework's "
```

visually hidden" utility whenever possible. You can
[read more about why `aria-label` might not be the best solution](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/).

> 

Reviewer: ```suggestion

### Example - Radix UI

https://www.radix-ui.com/primitives/docs/utilities/accessible-icon

***tsx
import { ArrowRightIcon } from 'lucide-react';
import { AccessibleIcon } from '@radix-ui/react-accessible-icon';

<AccessibleIcon label="Next item">
  <ArrowRightIcon />
</AccessibleIcon>
***
```

Author: https://www.radix-ui.com/primitives/docs/utilities/accessible-icon

Ought to be a link with descriptive text.
````suggestion
### Example - Radix UI

Use [Radix UI's built-in accessible icon utility component](https://www.radix-ui.com/primitives/docs/utilities/accessible-icon).

```tsx
import { ArrowRightIcon } from 'lucide-react';
import { AccessibleIcon } from '@radix-ui/react-accessible-icon';

<AccessibleIcon label="Next item">
  <ArrowRightIcon />
</AccessibleIcon>
```
````

> E.g. in case of Bootstrap:

Reviewer: ```suggestion
### Example - Bootstrap
```

```html

<div>
  <i data-lucide="phone" aria-hidden="true"></i>
  <span class="visually-hidden">Phone number</span>
</div>
```

> Or with tailwindcss:

Reviewer: ```suggestion
### Example - Tailwind CSS
```

![](../../images/a11y/alttext-iconbuttons.svg?raw=true)

## A note on `aria-label`

Although you could provide accessible labels to your elements via the `aria-label` attribute, we
generally suggest using your chosen CSS framework's "visually hidden" utility instead whenever
possible, since some assistive technology fails to announce `aria-label` attributes in case the

> element is not of an appropriate semantic role.

Reviewer: The main reason I don't like `aria-label` is that it gets exposed by all major browsers as AXDescription while normal content is exposed as AXTitle, and according to the spec, AXDescription should be read out in the system language, meaning your `lang=…` attribute gets ignored and your screen reader might sound funky.

Author: Neato, I've updated this description and removed all traces of `aria-label` from the illustrations.
